src/renderer/viz/expressions/top.js

```diff
@@ -59,56 +67,87 @@ export default class Top extends BaseExpression {
         this._textureBuckets = null;
     }
     get numCategories () {
-        return Math.round(this.buckets.eval()) + 1;
+        return this.numBuckets + 1;
+    }
+    get numBuckets () {
+        let buckets = Math.round(this.buckets.eval());
+        if (buckets > this.property.numCategories) {
+            buckets = this.property.numCategories;
+        }
+        if (buckets > MAX_TOP_BUCKETS) {
+            // setTimeout is used here because throwing within the renderer stack leaves the state in an invalid state,
+            // making this error an unrecoverable error, within the setTimeout the error is recoverable
+            const prev = this.buckets.eval();
+            setTimeout(() => {
+                throw new Error(`top() function has a limit of ${MAX_TOP_BUCKETS} buckets but '${prev}' buckets were specified`);
+            });
+            buckets = 0;
+        }
+        return buckets;
     }
     _applyToShaderSource (getGLSLforProperty) {
-        const property = this.property._applyToShaderSource(getGLSLforProperty);
+        const childSources = {};
+        this.childrenNames.forEach(name => { childSources[name] = this[name]._applyToShaderSource(getGLSLforProperty); });
         return {
-            preface: this._prefaceCode(property.preface + `uniform sampler2D topMap${this._uid};\n`),
-            inline: `(255.*texture2D(topMap${this._uid}, vec2(${property.inline}/1024., 0.5)).a)`
+            preface: this._prefaceCode(Object.values(childSources).map(s => s.preface).reduce((a, b) => a + b, '') + `
```

Comment:
You can use `.join('')` instead of `.reduce((a, b) => a + b, '')`

Reply:
:+1: